stdlib/mexpr/parser.mc

```diff
@@ -0,0 +1,433 @@
+    recursive
+    let remove = lam p. lam str.
+      match str with "\n" ++ xs then eatWSAC (advanceRow p 1) xs else
+      match str with [x] ++ xs then remove (advanceCol p 1) xs else
+      eatWSAC p str
+    in remove p xs
+end
+
+-- Eat multiline comment of the form *-  -*
+lang MultilineCommentParser = WSACParser
+  sem eatWSAC (p : Pos) =
+  | "/-" ++ xs ->
+    recursive
+    let remove = lam p. lam str. lam d.
+      match str with "/-" ++ xs then remove (advanceCol p 2) xs (addi d 1) else
+      match str with "\n" ++ xs then remove (advanceRow p 1) xs d else
+      match str with "-/" ++ xs then
+        if eqi d 1 then eatWSAC (advanceCol p 2) xs
+        else remove (advanceCol p 2) xs (subi d 1) else
+      match str with [_] ++ xs then remove (advanceCol p 1) xs d else
+      if eqi d 0 then eatWSAC p str else posErrorExit p "Unmatched multiline comments."
+    in remove (advanceCol p 2) xs 1
+end
+
+-- Commbined WSAC parser for MExpr
+lang MExprWSACParser = WhitespaceParser + LineCommentParser + MultilineCommentParser
+
+let _ = use MExprWSACParser in
+  utest eatWSAC (initPos "") " --foo \n  bar " with {str = "bar ", pos = posVal "" 2 2} in
+  utest eatWSAC (initPos "") " /- foo -/ bar" with {str = "bar", pos = posVal "" 1 11} in
+  utest eatWSAC (initPos "") " /- foo\n x \n -/ \nbar " with {str = "bar ", pos = posVal "" 4 0} in
+  utest eatWSAC (initPos "") " /- x -- y /- foo \n -/ -/ !" with {str = "!", pos = posVal "" 2 7} in
+  ()
+
+
+
+-- Top of the expression parser. Connects WSAC with parsing of other non terminals
+lang ExprParser = WSACParser
+  sem parseExpr (p: Pos) =
+  | s ->
+    let r1 = parseExprMain p 0 s in
+    let r2 = eatWSAC r1.pos r1.str in
+    if eqi (length r2.str) 0 then r1.val
+    else posErrorExit r2.pos "Parse error. Unknown characters."
+  
+  sem parseExprMain (p: Pos) (prec: Int) =
+  | s ->
+    let r1 = eatWSAC p s in
+    let exp = parseExprImp r1.pos r1.str in
+    let r2 = eatWSAC exp.pos exp.str in
+    parseInfix r2.pos prec exp r2.str 
+
+  sem parseInfix (p: Pos) (prec: Int) (exp: Expr) =
+
+  sem parseExprImp (p: Pos) =
+  | _ -> posErrorExit p "Parse error. Unknown character sequence."
+end
+
+
+-- Include this fragment is there are no infix operations
+lang ExprParserNoInfix = ExprParser
+  sem parseInfix (p: Pos) (prec: Int) (exp: Expr) =
+  | _ -> exp
+end
+
+-- Parsing of boolean literals
+lang BoolParser = ExprParser + ConstAst + BoolAst
+  sem parseExprImp (p: Pos) =
+  | "true" ++ xs ->
+      let p2 = advanceCol p 4 in
+      {val = TmConst {val = CBool {val = true}, fi = makeInfo p p2}, pos = p2, str = xs}
+  | "false" ++ xs ->
+      let p2 = advanceCol p 5 in
+      {val = TmConst {val = CBool {val = false}, fi = makeInfo p p2}, pos = p2, str = xs}
+end
+
+
+-- Parsing of an unsigned integer
+lang UIntParser = ExprParser + ConstAst + IntAst
+  sem parseExprImp (p : Pos) =
+  | ("0" ++ s) & xs | ("1" ++ s) & xs | ("2" ++ s) & xs | ("3" ++ s) & xs | ("4" ++ s) & xs |
```

Comment:
This could be shortened to `(['0' | '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9'] ++ s) & xs`

Sidenote: I guess it might be nice to have ranges in patterns.

Reply:
Good point. I was not aware that we allowed OR patterns inside sequences. Also, I agree, it would be good if we add range patterns: both for simplicity when writing out this kind of pattern, and also for compilation: it might be more efficient to check for ranges.